src/GitVersionCore.Tests/StringFormatWithExtensionTests.cs

```diff
@@ -0,0 +1,75 @@
+﻿using System;
+
+using GitVersion;
+using NUnit.Framework;
+
+namespace GitVersionCore.Tests
+{
+    [TestFixture]
+
+    public class StringFormatWithExtensionTests
+    {
+        [Test]
+        public void FormatWith_NoTokens()
+        {
+            var propertyObject = new { };
+            var target = "Some String without tokens";
+            var expected = target;
+            var actual = target.FormatWith(propertyObject);
+            Assert.AreEqual(expected, actual);
+        }
+
+        [Test]
+        public void FormatWith_SingleSimpleToken()
+        {
+            var propertyObject = new { SomeProperty = "SomeValue" };
+            var target = "{SomeProperty}";
+            var expected = "SomeValue";
+            var actual = target.FormatWith(propertyObject);
+            Assert.AreEqual(expected, actual);
+        }
+
+        [Test]
+        public void FormatWith_MultipleTokensAndVerbatimText()
+        {
+            var propertyObject = new { SomeProperty = "SomeValue", AnotherProperty = "Other Value" };
+            var target = "{SomeProperty} some text {AnotherProperty}";
+            var expected = "SomeValue some text Other Value";
+            var actual = target.FormatWith(propertyObject);
+            Assert.AreEqual(expected, actual);
+        }
+
+        [Test]
+        public void FormatWith_EnvVarToken()
+        {
+            Environment.SetEnvironmentVariable("GIT_VERSION_TEST_VAR", "Env Var Value");
+            var propertyObject = new { };
+            var target = "{$GIT_VERSION_TEST_VAR}";
```

Comment:
To avoid collisions, I would prefer if we could use a more explicit prefix here, like `env:`.

Reply:
Sure we could but I think from the reaction on approach suggested by @jbaehr on https://github.com/GitTools/GitVersion/issues/1157 we agreed to use $ as the prefix for enviroment variables and I am not sure as to what the collisions would be here since the usage of $ occurs in the format string and we parse them internally.

Reply:
The collision I'm thinking of is between non-environment variables and GitVersion's variables. We might introduce system variables and whatnot in the future, so having a prefix that clearly identifies where the variable is coming from is more future proof and intuitive than just `$` imho.

Reply:
Using `$` for dereferencing env vars was my initial proposal as for me it is very intuitive --
 it's was bash is using.
Anyway, I have no strong opinion on that; I'm also fine with `env:`.

Reply:
Addressed in the latest commit